### DISCUSSION TITLE
Week 12 Checkin

### DISCUSSION THREAD

**Discussion by NaNa7Miiii** [UW-Upwelling-Project/Summer-2023#17] [General]
#

**Comment by NaNa7Miiii** [comment]
Jiarui Monday Checkin:
* Trained a UNET model trying to answer "where the upwelling happens":

**Model Arch:**
```
def build_unet(input_shape):
    inputs = layers.Input(shape=input_shape)
    
    # encoder
    c1 = layers.Conv2D(64, (3, 3), activation='relu', padding='same')(inputs)
    p1 = layers.MaxPooling2D((2, 2))(c1)
    
    c2 = layers.Conv2D(128, (3, 3), activation='relu', padding='same')(p1)
    
    # decoder
    u2 = layers.Conv2DTranspose(64, (2, 2), strides=(2, 2), padding='same')(c2)
    
    # use padding
    u2 = layers.ZeroPadding2D(padding=((0, 1), (0, 1)))(u2)  # adjust the spatial dimensions
    
    c3 = layers.Conv2D(64, (3, 3), activation='relu', padding='same')(u2)
    
    outputs = layers.Conv2D(3, (1, 1), activation='softmax')(c3)
    
    return models.Model(inputs=inputs, outputs=outputs)
```
**Data:** 
Use our Zarr file, from 2000-01-01 to 2020-12-31, 60% - training, 20% - validation, 20% - test
**labels**
Here are the indexes for near-shore points:
<img width="486" alt="Screenshot 2023-09-04 at 7 09 44 PM" src="https://github.com/UW-Upwelling-Project/Summer-2023/assets/97627889/50071daa-858e-41b6-b0b4-229cf67f942d">
I previously paired each near-shore and off-shore point for each day and made a.csv file looks like:
<img width="882" alt="Screenshot 2023-09-04 at 7 12 30 PM" src="https://github.com/UW-Upwelling-Project/Summer-2023/assets/97627889/f05d5cd8-5595-4fa6-89dc-859b9c561141">
Then to label, I used Bresenham's line algorithm to draw lines between near-shore and off-shore points if (near shore sst - off shore sst < -1 which we would consider upwelling happens there). The way I label is land=0, non-upwelling regions=1, upwelling regions=2, and no negative labels just to make sure activation funds won't mess up. For example:
<img width="882" alt="Screenshot 2023-09-04 at 7 16 23 PM" src="https://github.com/UW-Upwelling-Project/Summer-2023/assets/97627889/bd8e3064-8288-44ff-b3f9-0672ad535b75">
For UNET in this case, if the input data for a given day has a shape of (149,181), then the label for that day must also has the same shape, so I made a xarray similar to Minh's xarray, it's just the sst values are now 0, 1, or 2.

**Model Effect:** 
<img width="977" alt="Screenshot 2023-09-04 at 1 05 27 PM" src="https://github.com/UW-Upwelling-Project/Summer-2023/assets/97627889/9928eb1e-ab39-4910-a9be-2327a7184f08">

This model can successfully show the main tracks, and due to the upwelling regions being the very minority of the whole picture, I am considering adjusting the weights for them. When I tried to increase the weight to 2, it turned out my weighting algorithm overreacted:
<img width="988" alt="Screenshot 2023-09-04 at 7 28 42 PM" src="https://github.com/UW-Upwelling-Project/Summer-2023/assets/97627889/fa5486e7-7bc4-4a7a-b755-8e98d410c8fc">

So my next task would be to seek a balance between those weights

**Reply by eeholmes** [reply]
Single day prediction

ValueError: This app has encountered an error. The original error message is redacted to prevent data leaks. Full error details have been recorded in the logs (if you're on Streamlit Cloud, click on 'Manage app' in the lower right of your app).
Traceback:
File "/home/appuser/venv/lib/python3.9/site-packages/streamlit/runtime/scriptrunner/script_runner.py", line 552, in _run_script
    exec(code, module.__dict__)
File "/app/upwelling_prediction/app.py", line 139, in <module>
    true_data = fetch_mur_data(start_time, end_time, slice(-5.0, 35.0), slice(45.0, 90.0))
File "/app/upwelling_prediction/utils.py", line 133, in fetch_mur_data
    resulting_dataset = xr.concat(sst_data_list, dim="time")
File "/home/appuser/venv/lib/python3.9/site-packages/xarray/core/concat.py", line 232, in concat
    raise ValueError("must supply at least one object to concatenate")

**Reply by eeholmes** [reply]
![image](https://github.com/UW-Upwelling-Project/Summer-2023/assets/2545978/4e8332cd-9160-4660-a6e8-12aa860ef164)

**Comment by NaNa7Miiii** [comment]
Jiarui Tuesday Checkin:
<img width="516" alt="Screenshot 2023-09-05 at 7 33 53 PM" src="https://github.com/UW-Upwelling-Project/Summer-2023/assets/97627889/a9aa3dbf-7239-4d2d-9a09-c6ed17deb0c7">
<img width="902" alt="Screenshot 2023-09-05 at 7 35 30 PM" src="https://github.com/UW-Upwelling-Project/Summer-2023/assets/97627889/65aaca2d-8467-4290-9390-de9514073954">

**Comment by eeholmes** [comment]
Text from ChatGPT

https://chat.openai.com/share/30040380-3d76-4d0a-ab24-da48f2b9a1ad

In this section, we detail the acquisition and integration of critical datasets from two renowned sources, the European Centre for Medium-Range Weather Forecasts (ECMWF) ERA5 reanalysis dataset and the Copernicus program. Our objective was to consolidate these datasets into a unified and efficient format, a Zarr file, for the purpose of comprehensive analysis. Specifically, we sought to extract essential parameters—sea surface temperature (SST), salinity, and sea surface height—from these sources to further our understanding of oceanic dynamics and their implications for climate studies. This section elucidates the methods employed to access, preprocess, and harmonize these datasets, facilitating subsequent analyses and insights into the complex interplay of Earth's oceans.

Zarr File Preparation

The creation of a Zarr file played a pivotal role in our data methodology due to its unique advantages for handling large-scale Earth datasets within the framework of cloud-native computing (add a source here. paper on the introduction to the Zarr format). In recent years, there has been a growing recognition in the remote-sensing community of the need for analysis-ready datasets (add source here on why cloud-native formats in the earth sciences). These are datasets that have been carefully processed and harmonized to remove inconsistencies, incorporate auxiliary information, and ensure data quality, making them readily usable for scientific analysis and modeling. <- edit this last sentence. *This is true in general but for us the main goal was to a) harmonize to a uniform 0.25 degree grid, b) harmonize to a daily dataset, c) pre-select sources for the environmental variables thus creating file that is ready for analyses in cloud-native workflows.*

Zarr, short for "zarr array," is a storage format specifically designed for efficient, scalable, and parallelizable access to multi-dimensional arrays, making it an ideal choice for managing Earth observation data (source, you can repeat the one from earlier or add a different one). This format leverages modern data storage technologies, such as chunked and compressed arrays, to optimize storage and retrieval, and reducing access latency. Chunking, an approach that divides large datasets into manageable blocks, is essential for accessing very large datasets, often in the terabyte and petabyte range, efficiently. Zarr, with its support for chunked storage, data compression, and lightweight metadata management, proves highly beneficial in this context. It optimizes storage, enables parallel processing, and minimizes access latency, all of which are critical when dealing with extensive datasets. Additionally, modern data storage technologies and strategies, including distributed storage systems and cloud-based solutions, play a crucial role in efficiently managing and accessing large datasets, ensuring that researchers can work with these data resources effectively.

By consolidating our ERA5 and Copernicus datasets into a Zarr file, we aimed to create an easy-to-use analysis-ready data cube, bringing together multiple diverse datasets onto the same grid for seamless integration. This preprocessing step not only harmonized the data but also ensured that it adhered to a consistent spatial and temporal framework, facilitating efficient and comprehensive analyses of Earth's oceans. This approach aligns with the broader shift in the remote-sensing community towards providing researchers with analysis-ready datasets, thus enabling more accessible and reproducible scientific investigations.

**Comment by eeholmes** [comment]
Google Scholar prompt: "comparison of the performance of machine learning algorithms for SST prediction"

https://www.sciencedirect.com/science/article/pii/S0952197622006650

Structure that might work. Results = "Experiments and discussion"
https://www.sciencedirect.com/science/article/pii/S0034425719303773?casa_token=oHeSF_V2Q5UAAAAA:QJ0sjcrn6vnRL0D3BOR-z9ok21P90qUMMtDSKvuHLKAD_sAyidrdzcD17WDK1ngod9LmMvNPUQ#s0025

Data
Methods
  Upwelling detection
  Machine learning (describe models and architecture)
Results and discussion
  Upwelling
Conclusions
  Talking about the bigger picture and model comparison
   Relate to other similar work. Did you find results that are similar to others.

**Comment by eeholmes** [comment]
Creating an outline for the results so readers can navigate.

Results

Upwelling detection

We summarize the performance of each model separately focusing on computation time metrics. We also discuss ideosyncracies for the models (problems and benefits). And then we compare all models.

4.1 ABC model

4.2 ZYZ modle

....

4.5 Comparison across all models

SST Prediction
